feat: add C ndarray interface and refactor implementation for `stats/base/dnanmeanpw`

aayush0325 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for `stats/base/dnanmeanpw`
-   refactors JavaScript implementation
-   updates tests, docs, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

No.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/src/main.c:
```diff
@@ -0,0 +1,52 @@
+#include "stdlib/blas/base/shared.h"
+#include "stdlib/blas/ext/base/dnannsumpw.h"
+#include "stdlib/strided/base/stride2offset.h"
+#include <stdint.h>
```
```suggestion
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/src/main.c:
```diff
@@ -0,0 +1,52 @@
+* @return         output value
+*/
+double API_SUFFIX(stdlib_strided_dnanmeanpw_ndarray)( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX ) {
+	int32_t n; // variable to store number of non `NaN` elements
+
+	const double sum = API_SUFFIX(stdlib_strided_dnannsumpw_ndarray)( N, X, strideX, offsetX, &n );
+	return sum / (double)n;
```
```suggestion
	CBLAS_INT n; // variable to store number of non `NaN` elements

	const double sum = API_SUFFIX(stdlib_strided_dnannsumpw_ndarray)( N, X, strideX, offsetX, &n );
	return sum / (double)n;
```

aayush0325 replied:
thanks for pointing that out!

headlessNode reviewed (commented):
The pointer should be `CBLAS_INT` type here.

See: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/blas/ext/base/dnannsumpw/README.md?plain=1#L242

EDIT: Looks like the JavaScript also needs to refactored.

aayush0325 commented:
> EDIT: Looks like the JavaScript also needs to refactored.

yes, i'm aware of that. i forgot to mark this as a draft :sweat_smile:

aayush0325 committed:
fix: use CBLAS_INT instead of int32_t

aayush0325 committed:
refactor: update js implementation and tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: failed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/lib/ndarray.js:
```diff
@@ -35,25 +31,22 @@ var WORKSPACE = [ 0.0, 0 ];
-	dnansumpw( N, WORKSPACE, x, stride, offset );
-	return WORKSPACE[ 0 ] / WORKSPACE[ 1 ];
+function dnanmeanpw( N, x, strideX, offsetX ) {
+	var out = new Float64Array( 2 );
+	var v = dnannsumpw( N, x, strideX, offsetX, out, 1, 0 );
+	return v[0] / v[1];
```
there's probably a bug here for negative strides

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/test/test.ndarray.js:
```diff
@@ -138,15 +136,13 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
 	t.end();
 });
 
 tape( 'the function supports a negative `stride` parameter', function test( t ) {
```
this is the test that is failing

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/lib/ndarray.js:
```diff
@@ -35,25 +31,22 @@ var WORKSPACE = [ 0.0, 0 ];
-	dnansumpw( N, WORKSPACE, x, stride, offset );
-	return WORKSPACE[ 0 ] / WORKSPACE[ 1 ];
+function dnanmeanpw( N, x, strideX, offsetX ) {
+	var out = new Float64Array( 2 );
+	var v = dnannsumpw( N, x, strideX, offsetX, out, 1, 0 );
+	return v[0] / v[1];
```
```suggestion
	var out = new Float64Array( 2 );
	dnannsumpw( N, x, strideX, offsetX, out, 1, 0 );
	return out[ 0 ] / out[ 1 ];
```

headlessNode replied:
`v` becomes redundant since we are passing `out`

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/test/test.ndarray.js:
```diff
@@ -162,9 +158,8 @@ tape( 'the function supports a negative `stride` parameter', function test( t )
-	N = floor( x.length / 2 );
 
-	v = dnanmeanpw( N, x, -2, 6 );
+	v = dnanmeanpw( 5, x, -2, 6 );
```
```suggestion
	v = dnanmeanpw( 5, x, -2, 8 );
```

headlessNode replied:
The `offset` for the negative case is incorrect.

kgryte replied:
@headlessNode Thanks for catching this! I believe I've seen this in other `nan*` stats packages, as well, so something to look out for @aayush0325.

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/test/test.ndarray.native.js:
```diff
@@ -171,9 +167,8 @@ tape( 'the function supports a negative `stride` parameter', opts, function test
-	N = floor( x.length / 2 );
 
-	v = dnanmeanpw( N, x, -2, 6 );
+	v = dnanmeanpw( 5, x, -2, 6 );
```
```suggestion
	v = dnanmeanpw( 5, x, -2, 8 );
```

headlessNode replied:
same comment.

aayush0325 committed:
fix: fix bugs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
refactor: update benchmarks, docs, examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: update readme

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -201,7 +293,7 @@ console.log( v );
 ## See Also
 
--   <span class="package-name">[`@stdlib/stats/base/dmeanpw`][@stdlib/stats/base/dmeanpw]</span><span class="delimiter">: </span><span class="description">calculate the arithmetic mean of a double-precision floating-point strided array using pairwise summation.</span>
+-   <span class="package-name">[`@stdlib/stats/base/dnanmeanpw`][@stdlib/stats/base/dnanmeanpw]</span><span class="delimiter">: </span><span class="description">calculate the arithmetic mean of a double-precision floating-point strided array using pairwise summation.</span>
```
this change isnt needed

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -222,7 +314,7 @@ console.log( v );
 <!-- <related-links> -->
 
-[@stdlib/stats/base/dmeanpw]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/dmeanpw
+[@stdlib/stats/base/dnanmeanpw]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/base/dnanmeanpw
```
same comment

aayush0325 committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: update examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -59,28 +59,25 @@ Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-p
 -   **N**: number of indexed elements.
 -   **x**: input [`Float64Array`][@stdlib/array/float64].
--   **stride**: index increment for `x`.
+-   **strideX**: stride length for `x`.
```
```suggestion
-   **strideX**: stride length.
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -59,28 +59,25 @@ Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-p
+-   **strideX**: stride length for `x`.
 
-The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in `x`,
+The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element in `x`,
```
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to compute the [arithmetic mean][arithmetic-mean] of every other element:
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -90,45 +87,39 @@ Note that indexing is relative to the first index. To introduce an offset, use [
 The function has the following additional parameters:
 
--   **offset**: starting index for `x`.
+-   **offsetX**: starting index for `x`.
```
```suggestion
-   **offsetX**: starting index.
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -90,45 +87,39 @@ Note that indexing is relative to the first index. To introduce an offset, use [
+-   **offsetX**: starting index for `x`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other element in `x` starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to calculate the [arithmetic mean][arithmetic-mean] for every other element starting from the second element:
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -181,6 +169,107 @@ console.log( v );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
```
```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -181,6 +169,107 @@ console.log( v );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
```
```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length.
-   **offsetX**: `[in] CBLAS_INT` starting index.
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/lib/ndarray.js:
```diff
@@ -35,25 +31,22 @@ var WORKSPACE = [ 0.0, 0 ];
+function dnanmeanpw( N, x, strideX, offsetX ) {
+	var out = new Float64Array( 2 );
+	dnannsumpw( N, x, strideX, offsetX, out, 1, 0 );
+	return out[0] / out[1];
```
```suggestion
	return out[ 0 ] / out[ 1 ];
```

headlessNode commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/src/main.c:
```diff
@@ -0,0 +1,51 @@
+* @param N        number of indexed elements
+* @param X        input array
+* @param strideX  stride length
+* @param offsetX  starting index for X
```
```suggestion
* @param offsetX  starting index
```

headlessNode reviewed (commented):
@aayush0325 A few documentation related suggestions.

1 - Use consistent terminology across similar packages and within the files of a single package. Notice that some files in this package use "stride length" and "starting index" instead of "stride length for x" and "starting index for x"

2 - For cases with multiple input arrays, we can then use specific terms like "stride length for x" and "stride length for y" to maintain clarity.

cc @kgryte

aayush0325 commented:
> @aayush0325 A few documentation related suggestions

That makes a lot of sense. We'd have to update the documentation for the existing packages as well. Dropping a patch soon!

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/lib/ndarray.js

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

aayush0325 committed:
Update lib/node_modules/@stdlib/stats/base/dnanmeanpw/src/main.c

Co-authored-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>
Signed-off-by: Aayush Khanna <96649223+aayush0325@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -59,28 +59,25 @@ Computes the [arithmetic mean][arithmetic-mean] of a double-precision floating-p
-var N = floor( x.length / 2 );
 
-var v = dnanmeanpw( N, x, 2 );
+var v = dnanmeanpw( 4, x, 2 );
```
```suggestion
var v = dnanmeanpw( 5, x, 2 );
```

While `4` is correct, it is not the intent. The point is to have a `NaN` value in the calculation. If `N = 4`, then computation stops at `4.0`. Probably worthwhile looking over the examples once more to sanity check.

kgryte replied:
It is this sort of thing which motivates why we've been removing the use of `floor`, as it can mask documentation and test bugs. Better to be explicit about the number of elements over which to perform computation.

aayush0325 committed:
chore: fix readme tests

aayush0325 committed:
docs: update examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
kindly take a look at the examples here, we'd have to update this in existing packages as well, dropping a patch for that soon!

kgryte commented:
@aayush0325 README examples LGTM

aayush0325 commented:
> @aayush0325 README examples LGTM

few other places where we had to make similar changes as well, see [this commit](https://github.com/stdlib-js/stdlib/pull/4763/commits/756db93c96f55aa61f9e45408c93159d93a49bb2), i will then make similar changes across existing packages

kgryte commented:
> few other places where we had to make similar changes as well

Those look fine to me, as well.

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/README.md:
```diff
@@ -45,42 +45,41 @@ The [arithmetic mean][arithmetic-mean] is defined as
 
 <section class="usage">
 
+<!-- lint disable maximum-heading-length -->
+
```
```suggestion
```

I don't believe this is necessary for this package.

kgryte committed:
style: enable lint rule

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
style: disable ESLint rules

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dnanmeanpw/lib/ndarray.js:
```diff
@@ -35,25 +31,22 @@ var WORKSPACE = [ 0.0, 0 ];
-	dnansumpw( N, WORKSPACE, x, stride, offset );
-	return WORKSPACE[ 0 ] / WORKSPACE[ 1 ];
+function dnanmeanpw( N, x, strideX, offsetX ) {
+	var out = new Float64Array( 2 );
```
@aayush0325 The `WORKSPACE` idea is still valid here. Namely, we can avoid allocating a new `out` array for each call to `dnanmeanpw` and just reuse the same workspace each time. While not thread safe in C, this is perfectly fine in JS.

aayush0325 replied:
i'll keep this in mind!

kgryte committed:
refactor: avoid repeated temporary array allocation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
